### PR TITLE
Add support for Control Flow Guard on Windows.

### PR DESCRIPTION
**Pull request by ajpaverd**
LLVM now supports Windows Control Flow Guard (CFG): https://github.com/llvm/llvm-project/commit/d157a9bc8ba1085cc4808c6941412322a7fd884e
This patch adds support for rustc to emit the required LLVM module flags to enable CFG metadata (cfguard=1) or metadata and checks (cfguard=2). The LLVM module flags are ignored on unsupported targets and operating systems.

### CODE REVIEW

File: src/librustc_codegen_ssa/back/linker.rs
Comment (jonas-schievink) [inline src/librustc_codegen_ssa/back/linker.rs]: I think it would be better to report an error instead of silently ignoring this

File: src/librustc_session/options.rs
Comment (jonas-schievink) [inline src/librustc_session/options.rs]: Why not make this a boolean option like `-Cno-redzone`?

File: src/librustc_session/options.rs
Comment (jonas-schievink) [inline src/librustc_session/options.rs]: This should probably start out as an unstable `-Z` flag (in `DebuggingOptions` below) instead of an insta-stable `-C` flag

File: src/librustc_codegen_ssa/back/linker.rs
Comment (ajpaverd) [reply src/librustc_codegen_ssa/back/linker.rs]: Thanks for the suggestion! Would it be sufficient to give a warning in each of these cases or should this be a fatal error?

File: src/librustc_session/options.rs
Comment (ajpaverd) [reply src/librustc_session/options.rs]: I don't think we can make this a boolean option because we need to encode three possibilities:
- option not present => do nothing
- `control_flow_guard=nochecks` => emit CFG metadata but no checks (corresponds to the Clang option `-cfguard,nochecks`). This is sometimes needed when linking against other binaries that expect CFG metadata.
- `control_flow_guard=checks` => emit CFG metadata and checks (corresponds to the Clang option `-cfguard`)

File: src/librustc_codegen_ssa/back/linker.rs
Comment (jonas-schievink) [reply src/librustc_codegen_ssa/back/linker.rs]: Not sure, I'll let the compiler team decide

File: src/librustc_session/options.rs
Comment (jonas-schievink) [reply src/librustc_session/options.rs]: Ah, I see, thanks for explaining!

File: src/librustc_codegen_ssa/back/linker.rs
Comment (bjorn3) [inline src/librustc_codegen_ssa/back/linker.rs]: It doesn't make sense for emscripten to ever support this.

File: src/librustc_codegen_ssa/back/linker.rs
Comment (bjorn3) [inline src/librustc_codegen_ssa/back/linker.rs]: It is unlikely that ptx will ever support this.

File: src/librustc_codegen_ssa/back/linker.rs
Comment (bjorn3) [inline src/librustc_codegen_ssa/back/linker.rs]: Wasm won't ever support it.

File: src/librustc_session/config.rs
Comment (bjorn3) [inline src/librustc_session/config.rs]: Please use doc comments instead.

**Review by bjorn3 — CHANGES_REQUESTED** [review]
Please remove `yet` from `not yet supported`.

File: src/librustc_session/config.rs
Comment (ajpaverd) [reply src/librustc_session/config.rs]: Thanks for the feedback!

File: src/librustc_codegen_ssa/back/linker.rs
Comment (ajpaverd) [reply src/librustc_codegen_ssa/back/linker.rs]: Thanks - is there something I should do to ask for further review?

Comment (jonas-schievink) [reply src/librustc_codegen_ssa/back/linker.rs]: No, it just might take a while until the reviewer can get to it

File: src/librustc_session/config.rs
Comment (nagisa) [inline src/librustc_session/config.rs]: Please rename this variant to `MetadataOnly` or something else along the lines. `Nochecks` seems to imply CFG is disabled entirely, which it isn’t.

File: src/librustc_session/options.rs
Comment (nagisa) [inline src/librustc_session/options.rs]: It is typical for these kinds options to also accept a value that explicitly disables the behaviour, This option should have such a value as well. In doing so you should also add an explicit `Disabled` or `No` variant to `CFGuard` (and store just `CFGuard` rather than `Option<CFGuard>`).

File: src/librustc_session/options.rs
Comment (nagisa) [inline src/librustc_session/options.rs]: Nit: `slote` -> `slot`.

File: src/librustc_session/options.rs
Comment (nagisa) [inline src/librustc_session/options.rs]: Nit: a better place for this comment would be on the `CFGuard` type or on the function that parses it, rather than here.

**Review by nagisa — COMMENTED** [review]
r=me sans some cosmetic nits.

File: src/librustc_session/options.rs
Comment (ajpaverd) [reply src/librustc_session/options.rs]: Fixed

File: src/librustc_session/options.rs
Comment (ajpaverd) [reply src/librustc_session/options.rs]: Fixed

File: src/librustc_session/options.rs
Comment (ajpaverd) [reply src/librustc_session/options.rs]: Thanks for the suggestion! I've added a `CFGuard::Disabled` variant, and am now just storing `CFGuard`.

File: src/librustc_session/config.rs
Comment (ajpaverd) [reply src/librustc_session/config.rs]: The `nochecks` variant was intended to match the corresponding `-cfguard,nochecks` option in Clang (https://reviews.llvm.org/rL339420). Should we keep this terminology for consistency?

The intention is really to convey that this is "Control Flow Guard without any run-time checks". Is there somewhere in the documentation I could explain this?

Comment (nagisa) [reply src/librustc_session/config.rs]: I guess you can keep it in that case. Just `s/Nochecks/NoChecks/`. There are a couple of places for this documentation to go, including in the manpages and the [unstable book](https://doc.rust-lang.org/stable/unstable-book/compiler-flags.html).

File: src/librustc_codegen_llvm/context.rs
Comment (nagisa) [inline src/librustc_codegen_llvm/context.rs]: Make this match exhaustive (i.e. `s/_/CFGuard::Disabled/`)

File: src/librustc_session/config.rs
Comment (nagisa) [inline src/librustc_session/config.rs]: Nit: You could comment each variant separately.

Comment (ajpaverd) [reply src/librustc_session/config.rs]: Fixed

File: src/librustc_codegen_llvm/context.rs
Comment (ajpaverd) [reply src/librustc_codegen_llvm/context.rs]: Fixed

**Comment by nagisa** [comment]
@bors r+

Thanks!

**Comment by ajpaverd** [comment]
> @bors r+
> 
> Thanks!

Thanks @nagisa! Quick question: should I squash the commits in this PR and rebase onto master to avoid including the merge commit?

**Comment by nagisa** [comment]
Sure, please do squash.

**Comment by ajpaverd** [comment]
> Sure, please do squash.

Did my squash+rebase inadvertently interrupt bors' testing and merging?

**Comment by bjorn3** [comment]
> Did my squash+rebase inadvertently interrupt bors' testing and merging?

Yes, any push will prevent a PR from being merged before someone approves it again.

**Comment by ajpaverd** [comment]
Is there a way to figure out what caused the bors tests to time out (the logs just seem to say the operation was cancelled)?

**Comment by nagisa** [comment]
Looks like it may have been cancelled manually (https://dev.azure.com/rust-lang/rust/_build/results?buildId=19568&view=results) probably to give way to something more important. Lets just retry for now. Thanks for bringing it to my attention.

@bors r+

**Pull request merged**
